substrate/frame/election-provider-multi-block/src/verifier/impls.rs

```diff
@@ -0,0 +1,680 @@
+			&all_voter_pages,
+			&targets,
+			solution,
+			desired_targets,
+			page,
+		)
+	}
+}
+
+impl<T: impls::pallet::Config> AsyncVerifier for Pallet<T> {
+	type SolutionDataProvider = T::SolutionDataProvider;
+
+	fn force_finalize_verification(claimed_score: ElectionScore) -> Result<(), FeasibilityError> {
+		Self::finalize_async_verification(claimed_score)
+	}
+
+	fn status() -> Status {
+		VerificationStatus::<T>::get()
+	}
+
+	fn start() -> Result<(), &'static str> {
+		if let Status::Nothing = Self::status() {
+			let claimed_score = Self::SolutionDataProvider::get_score().unwrap_or_default();
+
+			if Self::ensure_score_quality(claimed_score).is_err() {
+				Self::deposit_event(Event::<T>::VerificationFailed(
+					crate::Pallet::<T>::msp(),
+					FeasibilityError::ScoreTooLow,
+				));
+				// report to the solution data provider that the page verification failed.
+				T::SolutionDataProvider::report_result(VerificationResult::Rejected);
+				// despite the verification failed, this was a successful `start` operation.
+				Ok(())
+			} else {
+				VerificationStatus::<T>::put(Status::Ongoing(crate::Pallet::<T>::msp()));
+				Ok(())
+			}
+		} else {
+			sublog!(warn, "verifier", "tries to start election while ongoing, ignored.");
+			Err("verification ongoing")
+		}
+	}
+
+	fn stop() {
+		sublog!(warn, "verifier", "stop signal received. clearing everything.");
+		QueuedSolution::<T>::clear_invalid_and_backings();
+
+		// if a verification is ongoing, signal the solution rejection to the solution data
+		// provider and reset the current status.
+		VerificationStatus::<T>::mutate(|status| {
+			if matches!(status, Status::Ongoing(_)) {
+				T::SolutionDataProvider::report_result(VerificationResult::Rejected);
+			};
+			*status = Status::Nothing;
+		});
+	}
+
+	// Sets current verifications status.
+	#[cfg(any(test, feature = "runtime-benchmarks"))]
+	fn set_status(status: Status) {
+		VerificationStatus::<T>::put(status);
+	}
+}
+
+impl<T: impls::pallet::Config> Pallet<T> {
+	fn do_on_initialize(_now: crate::BlockNumberFor<T>) -> Weight {
+		let max_backers_winner = T::MaxBackersPerWinner::get();
+		let max_winners_page = T::MaxWinnersPerPage::get();
+
+		match crate::Pallet::<T>::current_phase() {
+			// reset remaining unsigned pages after snapshot is created.
+			crate::Phase::Snapshot(page) if page == crate::Pallet::<T>::lsp() => {
+				RemainingUnsignedPages::<T>::mutate(|remaining| {
+					*remaining = BoundedVec::truncate_from(
+						(crate::Pallet::<T>::lsp()..crate::Pallet::<T>::msp() + 1)
+							.collect::<Vec<_>>(),
+					);
+				});
+
+				sublog!(
+					debug,
+					"verifier",
+					"reset remaining unsgined pages to {:?}",
+					RemainingUnsignedPages::<T>::get()
+				);
+			},
+			_ => (),
+		}
+
+		if let Status::Ongoing(current_page) = <VerificationStatus<T>>::get() {
+			let maybe_page_solution =
+				<T::SolutionDataProvider as SolutionDataProvider>::get_paged_solution(current_page);
```

Comment:
`get_paged_solution` returns the leader's solution so if it's none (no leader solution) `T::SolutionDataProvider::report_result` will panic because it does `Submissions::<T>::take_leader_score` which already is none. 
So maybe `report_result` shouldn't check for the leader's score or there is some other change needed in the logic.